QccPack-0.61-1/src/libQccPackWAV/utils/man1/tceencode3d.1
.TH TCEENCODE3D 1 "QCCPACK" ""
.SH NAME
tceencode3d, tcedecode3d \-
encode/decode an image cube using the 3D-TCE algorithm
.SH SYNOPSIS
.sp
.B tceencode3d
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-nl
.IR  num_levels "\|]"
.RB "[\|" \-sl
.IR  spatial_num_levels "\|]"
.RB "[\|" \-tl
.IR  temporal_num_levels "\|]"
.RB "[\|" \-a
.IR  alpha "\|]"
.RB "[\|" \-vo "\|]"
.I rate
.I icbfile
.I bitstream
.LP
.B tcedecode3d
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-r
.IR  rate "\|]"
.I bitstream
.I icbfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String. 
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP 
.BI \-nl " num_levels"
Integer. Number of levels of dyadic decomposition; default value is 5 levels.
.TP 
.BI \-sl " spatial_num_levels"
Integer. Number of levels of spatial decomposition; no default value.
.TP 
.BI \-tl " temporal_num_levels"
Integer. Number of levels of temporal decomposition; no default value.
.TP
.BI \-a " alpha"
Float. Value of the learning-rate parameter alpha; default is 0.2.
.TP
.BI \-r " rate"
Float. Rate at which to stop decoding. Default is to decode until
end of input bitstream is reached.
.TP
.B \-vo
Print only the value of the rate to stdout (terse output);
default is verbose output.
.SH DESCRIPTION
.LP
.B tceencode3d
codes an image cube,
.IR icbfile ,
using the 3D-TCE algorithm.
The original TCE algorithm was developed for 2D images by
Tian and Hemami;
it was latter extended to 3D by Zhang
.IR "et al" .
The TCE (tarp coding using classification to achieve embedding) algorithm
is based on the tarp algorithm (see 
.BR QccWAVTarpEncode (3)),
but is designed to provide better rate-distortion performance
when used in an embedded fashion; i.e., when decoding is performed
at a rate less than that at which the bitstream was produced.
See 
.BR QccWAVtce3DEncode (3)
for more detail on the TCE algorithm.
.LP
The transform is the usual discrete wavelet transform (DWT),
with the
.B \-w
option specifying the wavelet, and the
.B \-b
option specifying the boundary extension.
.LP
The 3D-TCE algorithm performance is determined in part through
the parameter
.IR alpha ,
a value that gives the learning rate of the density-estimation
process implemented by the tarp filter used in one of
the coding passes of the TCE algorithm.
.LP
The
.BR \-nl ,
.BR \-sl ,
and
.BR \-tl
options determine whether a dyadic or wavelet-packet transform is used.
If
.B \-nl
is specified, a dyadic transform is used. In this case,
.IR num_levels
is the number of decompositions in all three directions.
If either
.B \-sl
or
.BR \-tl ,
or
both, is specified, a wavelet-packet transform is used.
In this case,
.I spatial_num_levels
and
.I temporal_num_levels
give the number of decompositions spatially and temporally, respectively,
of the packet transform. If only one of
.B \-sl
and
.B \-tl
is given, then the same number of decompositions
is used spatially and temporally.
In the case that none of
.BR \-nl ,
.BR \-sl ,
and
.BR \-tl 
is specified, then a dyadic decomposition with the default value of
.I num_levels
is used.
.LP
The desired bitrate of the output
.I bitstream
is specified by
.IR rate ,
which is expressed in bits per voxel of the original image.
.B tceencode3d
normally prints to stdout a small amount of information
concerning this target bitrate and the actual bitrate achieved.
Usually these rates are identical, but may differ slightly.
If the 
.B \-vo
(value-only) option is given, all printed output is suppressed except for
the actual bitrate.
.LP
.B tcedecode3d
performs the corresponding decoding using the 3D-TCE algorithm, decoding
the
.I bitstream
made by
.B tceencode3d
and producing
a reconstructed image cube,
.IR icbfile .
For sane results, 
.IR wavelet 
and
.IR boundary 
should be the same as they were for
.BR tceencode3d .
If a decoding rate is specified with the
.B \-r
option, then
.B tcedecode3d
stops decoding when
.I rate
bits per voxel from the input bitstream
have been decoded. Otherwise, decoding stops when the
end of the input bitstream is reached.
.LP
.BR tceencode3d
and
.BR tcedecode3d
call
.BR QccWAVtce3DEncode (3)
and
.BR QccWAVtce3DDecode (3),
respectively, to perform 3D-TCE encoding and decoding.
Refer to the man pages for these functions for more information
on the details of the QccPack implementation of the 3D-TCE algorithm.
.SH "SEE ALSO"
.BR QccWAVtce3DEncode (3),
.BR QccWAVtce3DDecode (3),
.BR QccWAVSubbandPyramid3D (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

.LP
C. Tian and S. S. Hemami, "An Embedded Image Coding System
Based on Tarp Filter with Classification," in
.IR "Proceedings of the International Conference on Acoustics, Speech, and Signal Processing" ,
Montreal, Quebec, Canada, May 2004, vol. 3, pp. 49-52.

J. Zhang, J. E. Fowler, and G. Liu,
"Lossy-to-Lossless Compression of Hyperspectral Imagery Using
3D-TCE and an Integer KLT," 
.IR "IEEE Geoscience and Remote Sensing Letters" ,
April 2008, submitted.

E. Ordentlich, M. Weinberger, and G. Seroussi,
"A Low-Complexity Modeling Approach for Embedded Coding of
Wavelet Coefficients," in
.IR "Proceedings of the IEEE Data Compression Conference" ,
Snowbird, UT, March 2002, pp. 23-32.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.




QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/tceencode3d.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


/* 
 *
 * Written by
 *
 * Jing Zhang, at Mississippi State University, 2008 
 *
 */

#include "tceencode3d.h"

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-nl %: %d:num_levels] [-sl %: %d:spatial_num_levels] [-tl %: %d:temporal_num_levels] [-a %f:alpha]  [-vo %:] %f:rate %s:icbfile %s:bitstream"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

int TransformType = QCCWAVSUBBANDPYRAMID3D_DYADIC;
int NumLevelsSpecified = 0;
int NumLevels = 5;
int SpatialNumLevelsSpecified = 0;
int SpatialNumLevels = 0;
int TemporalNumLevelsSpecified = 0;
int TemporalNumLevels = 0;

float Alpha = 0.2;

QccIMGImageCube InputImage;
int ImageNumFrames, ImageNumRows, ImageNumCols;

QccBitBuffer OutputBuffer;

int ValueOnly = 0;

float TargetRate;
float ActualRate;
int NumPixels;
int TargetBitCnt;


int main(int argc, char *argv[])
{
  QccInit(argc, argv);
  
  QccWAVWaveletInitialize(&Wavelet);
  QccIMGImageCubeInitialize(&InputImage);
  QccBitBufferInitialize(&OutputBuffer);
  
  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &NumLevelsSpecified,
                         &NumLevels,
                         &SpatialNumLevelsSpecified,
                         &SpatialNumLevels,
                         &TemporalNumLevelsSpecified,
                         &TemporalNumLevels,
                         &Alpha,                    
                         &ValueOnly,
                         &TargetRate,
                         InputImage.filename,
                         OutputBuffer.filename))
    QccErrorExit();
  
  if ((NumLevels < 0) || (SpatialNumLevels < 0) || (TemporalNumLevels < 0))
    {
      QccErrorAddMessage("%s: Number of levels of decomposition must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  if (NumLevelsSpecified &&
      (SpatialNumLevelsSpecified || TemporalNumLevelsSpecified))
    {
      QccErrorAddMessage("%s: If num_levels is given (dyadic transform), neither spatial_num_levels or temporal_num_levels (packet transform) can be specified",
                         argv[0]);
      QccErrorExit();
    }
  
  if (NumLevelsSpecified ||
      ((!SpatialNumLevelsSpecified) && (!TemporalNumLevelsSpecified)))
    {
      TransformType = QCCWAVSUBBANDPYRAMID3D_DYADIC;
      SpatialNumLevels = NumLevels;
      TemporalNumLevels = NumLevels;
    }
  else
    {
      TransformType = QCCWAVSUBBANDPYRAMID3D_PACKET;
      if (SpatialNumLevelsSpecified && (!TemporalNumLevelsSpecified))
        TemporalNumLevels = SpatialNumLevels;
      else
        if ((!SpatialNumLevelsSpecified) && TemporalNumLevelsSpecified)
          SpatialNumLevels = TemporalNumLevels;
    }
  
  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()", argv[0]);
      QccErrorExit();
    }
  
  if (QccIMGImageCubeRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageCubeRead()",
                         argv[0]);
      QccErrorExit();
    }
  ImageNumFrames = InputImage.num_frames;
  ImageNumRows = InputImage.num_rows;
  ImageNumCols = InputImage.num_cols;
  
  NumPixels = ImageNumFrames * ImageNumRows * ImageNumCols;
  
  OutputBuffer.type = QCCBITBUFFER_OUTPUT;
  if (QccBitBufferStart(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferStart()",
                         argv[0]);
      QccErrorExit();
    }
  
  TargetBitCnt = (int)(ceil((NumPixels * TargetRate)/8.0))*8;
  
  if (QccWAVtce3DEncode(&(InputImage),                       
                        &OutputBuffer,
                        TransformType,
                        TemporalNumLevels,
                        SpatialNumLevels,
                        Alpha,
                        &Wavelet,
                        TargetBitCnt))
    {
      QccErrorAddMessage("%s: Error calling QccWAVtce3DEncode()",
                         argv[0]);
      QccErrorExit();
    }
  
  ActualRate = (double)OutputBuffer.bit_cnt / NumPixels;
  
  if (QccBitBufferEnd(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferEnd()", argv[0]);
      QccErrorExit();
    }
  
  if (ValueOnly)
    printf("%f\n", ActualRate);
  else
    {
      printf("3D-TCE coding of %s:\n", InputImage.filename);
      printf("  Target rate: %f bpv\n", TargetRate);
      printf("  Actual rate: %f bpv\n", ActualRate);
    }
  
  QccIMGImageCubeFree(&InputImage); 
  QccWAVWaveletFree(&Wavelet);
  
  QccExit;
}
